### PR TITLE
Allow user to get a Post

### PR DESCRIPTION
**Pull request by thuyvy3110**
we need to fetch the full a post #31

### CODE REVIEW

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostExceptionController.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.dawn.jat.illuminati.post.controller;
+
+import com.dawn.jat.illuminati.post.PostNotfoundException;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/controller/PostExceptionController.java]: Should put it in src/main/java/com/dawn/jat/illuminati/post/advice

File: src/main/java/com/dawn/jat/illuminati/post/entity/PostEntity.java [hunk 2]
```diff
@@ -2,18 +2,25 @@
...
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
 import org.springframework.data.annotation.Id;
 import org.springframework.data.mongodb.core.mapping.Document;
 
 @Generated
+@RequiredArgsConstructor
 @Data
 @Document(collection = "post")
 public class PostEntity {
```
[the remark this answers is not in the document]
Comment (DeKal) [reply src/main/java/com/dawn/jat/illuminati/post/entity/PostEntity.java]: And we need full content of the post

File: src/main/java/com/dawn/jat/illuminati/post/PostNotfoundException.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package com.dawn.jat.illuminati.post;
+
+public class PostNotfoundException extends RuntimeException {
+    private static final long serialVersionUID = 1L;
```
[the remark this answers is not in the document]
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/PostNotfoundException.java]: delete serialVersionUID, it is not generated by Intellij

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 4]
```diff
@@ -14,10 +17,13 @@
...
 @RestController
-@RequestMapping({"/post"})
+@RequestMapping({"/api/post"})
 public class PostController {
     @Autowired
-    PostService postService;
+    private PostService postService;
+
+    @Autowired
+    private PostRepository postRepository;
```
[the remark this answers is not in the document]
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Remove postRepository

File: src/main/java/com/dawn/jat/illuminati/post/entity/PostEntity.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/entity/PostEntity.java]: Added author name, full post content

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 5]
```diff
@@ -27,15 +33,13 @@
     /**
      * Gets a Post by the Id.
      */
-    @RequestMapping(value = "/{id}")
-    public ResponseEntity<String> getPostById(@PathVariable("id") String id) {
-        Optional<PostEntity> post = postService.findById(id);
+    @GetMapping(value = "/{id}")
+    public ResponseEntity<Object> getAPostById(@PathVariable("id") String id) {
```
[the remark this answers is not in the document]
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Revert name getPostById

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostExceptionController.java (hunk 1 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostExceptionController.java]: Moved to advice folder

**Comment by DeKal** [comment]
We need to upload a sample DB for this PR

**Comment by DeKal** [comment]
PostExceptionController -> PostExceptionAdvice

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java [hunk 6]
```diff
@@ -27,15 +29,13 @@
...
-        Optional<PostEntity> post = postService.findById(id);
+    @GetMapping(value = "/{id}")
+    public ResponseEntity<Object> getPostById(@PathVariable("id") String id) {
+        Optional<PostEntity> postEntity = postService.findById(id);
 
-        if (post.isPresent()) {
-            return ResponseEntity.status(HttpStatus.OK)
-                    .body("OK");
+        if (!postEntity.isPresent()) {
+            throw new PostNotfoundException();
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: throw new PostNotfoundException("Cannot find post")
-> constructor (string msg)
-> getMessage return msg

File: src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionController.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class PostExceptionController {
+    @ExceptionHandler(value = PostNotfoundException.class)
+    public ResponseEntity<Object> exception(PostNotfoundException exception) {
+        return new ResponseEntity<>("Post not found", HttpStatus.NOT_FOUND);
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionController.java]: exception.getMessage()

File: src/main/java/com/dawn/jat/illuminati/post/entity/PostEntity.java [hunk 8]
```diff
@@ -2,18 +2,27 @@
...
     private String id;
-    private String title;
-    private String tag;
-    private String brief;
-    private String time;
+    @NonNull private String title;
+    @NonNull private String brief;
+    @NonNull private String time;
+    @NonNull private String[] tag;
+    @NonNull private String author;
```
Comment (DeKal) [inline src/main/java/com/dawn/jat/illuminati/post/entity/PostEntity.java]: add slug custom slug for users to find

File: src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java [hunk 9]
```diff
@@ -62,34 +58,24 @@ void findPostAll_whenNoRecord() {
 
     @Test
     void findPostAll_whenRecord() {
-        Mockito.when(postService.findAll()).thenReturn(Arrays.asList(postEntity1));
-        assertThat(postController.findAll().size(), is(1));
+        Mockito.when(postService.findAll()).thenReturn(Arrays.asList(postEntity1, postEntity2));
```
Comment (DeKal) [inline src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java]: separate this into 2 parts 
List<String> mockPostEntities = Arrays.asList(postEntity1, postEntity2);
Mockito.when(postService.findAll()).thenReturn(mockPostEntities)

File: src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java [hunk 10]
```diff
@@ -62,34 +58,24 @@ void findPostAll_whenNoRecord() {
...
-    @Test
-    public void postEntity_doNotHave_unexpectedValue() {
-        assertFalse(postEntity1.toString().contains("@$*&$abzn&*"));
-    }
-
     @Test
     public void whenPostIdIsAvail_thenRetrievedPostIsCorrect() {
-        Mockito.when(postService.findById("1")).thenReturn(Optional.of(new PostEntity()));
-        ResponseEntity<String> responseEntity = postController.getPostById("1");
+        Mockito.when(postService.findById("1")).thenReturn(Optional.of(postEntity1));
```
Comment (DeKal) [inline src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java]: mockPostEntity = Optional.of(postEntity1)

File: src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java (hunk 6 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/controller/PostController.java]: Updated

File: src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java [hunk 11]
```diff
@@ -62,34 +58,24 @@ void findPostAll_whenNoRecord() {
...
     }
 
     @Test
-    public void whenPostIdIsUnavail_thenRetrievedPostIsIncorrect() {
-        ResponseEntity<String> responseEntity = postController.getPostById("1A467");
-        assertEquals(HttpStatus.BAD_REQUEST, responseEntity.getStatusCode());
-        assertNotEquals(HttpStatus.OK, responseEntity.getStatusCode());
+    void whenPostIdIsUnavail_thenRetrievedPostIsIncorrect() {
+        assertThrows(PostNotfoundException.class, () -> {
+            postController.getPostById("1A467");
```
Comment (DeKal) [inline src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java]: nothing to make sure id is not existed 
try to make mock of controller
```
    PostController mockPostController = mock(PostController.class);
    Mockito.when(mockPostController.getPostById("1A467")).thenThrow(new PostNotfoundException(){});
        assertThrows(PostNotfoundException.class, () -> {
            mockPostController.getPostById("1A467");
        });
```

File: src/test/java/com/dawn/jat/illuminati/post/service/PostServiceTest.java [hunk 12]
```diff
@@ -60,8 +55,8 @@ public void findAllPost_whenRecord() {
 
     @Test
     public void findPostById() {
-        Mockito.when(postRepository.findById("1")).thenReturn(Optional.of(new PostEntity()));
-        assertThat(postService.findById("1"), is(Optional.of(new PostEntity())));
+        Mockito.when(postRepository.findById("1")).thenReturn(Optional.of(postEntity));
```
Comment (DeKal) [inline src/test/java/com/dawn/jat/illuminati/post/service/PostServiceTest.java]: same with other

File: src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java (hunk 11 above)
Comment (thuyvy3110) [reply src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java]: Updated

File: src/test/java/com/dawn/jat/illuminati/post/service/PostServiceTest.java (hunk 12 above)
Comment (thuyvy3110) [reply src/test/java/com/dawn/jat/illuminati/post/service/PostServiceTest.java]: Updated

File: src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java (hunk 10 above)
Comment (thuyvy3110) [reply src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java]: Updated

File: src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java (hunk 9 above)
Comment (thuyvy3110) [reply src/test/java/com/dawn/jat/illuminati/post/controller/PostControllerTest.java]: Updated

File: src/main/java/com/dawn/jat/illuminati/post/entity/PostEntity.java (hunk 8 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/entity/PostEntity.java]: image

Updated

File: src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionController.java (hunk 7 above)
Comment (thuyvy3110) [reply src/main/java/com/dawn/jat/illuminati/post/advice/PostExceptionController.java]: Updated

**Comment by DeKal** [comment]
LGTM 👍

**Pull request merged**